[language] Add changelog for phantom types

nilehmann opened the pull request:
Adds language changelog entry for phantom types

nilehmann commented on language/changes/5-phantom-type-params.md:
```diff
@@ -0,0 +1,77 @@
+# Phantom Type Parameters
+
+* Status: Implemented in Move 1.3
```
Not sure what should be the version

tnowacki replied:
I think it is 1.4

tnowacki reviewed (commented):
Could you try to format it to keep lines at 100 characters? It will make things a lot easier for reviewing :) 

If you are using vs code, you can install Prettier and set the width to "100" and the "Prose Wrap" setting to "always"

tnowacki reviewed (commented):
Looks pretty good, thanks for the writeup! 

Hopefully @bob-wilson can give it a once over before it needs to go in :)

nilehmann commented on language/changes/5-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+didn't make a difference on whether a type parameter was phantom, but since Move version 1.4 there
+is the possibility to explicitly declare a type parameters in a struct definition as phantom. A
+parameter declared as phantom is not considered when computing the
+[conditional abilities with generic types](3-abilities#Conditional-Abilities-with-Generic-Types).
```
It was broken, but I fixed it and is working here (https://github.com/diem/diem/blob/efd0152f068f80c7c76c7190a0d0bcb91ab9ed5f/language/changes/5-phantom-type-params.md)

nilehmann commented:
@tnowacki fixed your comments. Also hope @bob-wilson can take a look.

bob-wilson commented:
I'll try to look soon. I'll start by pointing out that the draft changelog for "named addresses" is number 5, so this one should be number 6.

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+
+## Motivations
+
+A phantom type parameter may come in handy in a couple of situations. For example, the struct
```
This first paragraph in the Motivations section seems misplaced. It says nothing about the motivation for phantom type parameters, but instead offers examples from the Diem Framework where they can be used. We should move this elsewhere. Since the other Motivations paragraph refers to these specific examples, I suggest moving this paragraph into the Introduction section.

A few other things:

- Let's also drop the first sentence. This is more substantial than providing something "handy".
- Replace "stdlib" with "the Move standard library"
- The term "notorious" livens up the tone of this document but seems too strong. Maybe "prominent"?

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+currency of the coin and allowing code to be written generically on any currency. As a last example,
+the capabilities in `Vault` also rely heavily on phantom type parameters.
+
+Previously, these examples required spurious ability annotations to satisfy the requirements of the
```
```suggestion
Previously, phantom type parameters required spurious ability annotations to satisfy the requirements of the
```

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+the types `XUS::XUS` and `XDX::XDX` were forced to have a `store` annotation even though they were
+never stored in the database! More over the extra `store` requirement polluted the entire Diem
+Framework and many functions had to specify an extra `store` constraints on their type parameters.
+With the new feature, `Diem::Diem<CoinType>` now declares `CoinType` to be phantom which avoids the
```
It would be better to first state this generally, i.e., saying that this language change avoids the need for spurious abilities, before giving the specific example of `Diem<CoinType>`. It might also be good to add something before this concluding sentence to elaborate on why the spurious abilities are bad.

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+
+The following two examples show valid uses of phantom parameters. In the first one, the parameter
+`T1` is not used at all inside the struct definition. In the second one, the parameter `T1` is only
+used as an argument to a parameter also declared as phantom.
```
I find this "declared as phantom" terminology awkward. I don't have a good suggestion for a direct replacement, but we could try to avoid using it more than necessary. For example, this sentence could say "...used as an argument to another phantom type parameter." Perhaps you could rewrite some of the other text with this terminology in a similar fashion?

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,103 @@
+
+Ability constraints and phantom type parameters are orthogonal features in the sense that phantom
+parameters can be declared with ability constraints. And any type parameter must satisfy that
+constraint, even thought it is phantom. For example, the following definition is perfectly valid:
```
A few nits on this sentence:
- typo: "thought" should be "though
- It's not a complete sentence. At a minimum, you need a comma after the initial "and", but probably there's a better way.
- It is not clear what "that constraint" refers to. The preceding sentence refers to "ability constraints" in general but says nothing about a specific constraint.
- It is also not clear what "it" refers to. Presumably "any type parameter", but it's confusing because it looks like it could be "that constraint" (which makes no sense).

nilehmann commented:
@bob-wilson I incorporated your comments, except maybe for this one

> It would be better to first state this generally, i.e., saying that this language change avoids the need for spurious abilities, before giving the specific example of Diem<CoinType>. It might also be good to add something before this concluding sentence to elaborate on why the spurious abilities are bad.

But the entire section has a different structure now that I moved some stuff to the intro, so it may not be too relevant anymore?

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,106 @@
+
+Ability constraints and phantom type parameters are orthogonal features in the sense that phantom
+parameters can be declared with ability constraints. When instantiating a phantom type parameter
+with an ability constraints, the type argument has to satisfy that constraint, even though the
```
```suggestion
with an ability constraint, the type argument has to satisfy that constraint, even though the
```

bob-wilson reviewed (changes requested):
I found one new minor typo, and I still think my previous comment applies. Here's what I think is missing from the current Motivations section:

- What's the problem with spurious abilities? Are they just a nuisance or are they more important than that?
- There is no direct statement that phantom types solves the problem in general. You explain that previously we needed spurious abilities and show some examples of that, and then conclude with a statement that one specific instance of the problem is now fixed. We should have a more general statement and then use that specific instance as an example.

bob-wilson commented on language/changes/6-phantom-type-params.md:
```diff
@@ -0,0 +1,112 @@
+
+Previously, defining a struct `S` with a phantom type parameter required spurious ability
+annotations to satisfy the requirements of the abilities declared for `S`. For a module author, this
+meant loosing control over structs that had to be declared with unnecessary abilities. Moreover, the
```
```suggestion
meant losing control over structs that had to be declared with unnecessary abilities. Moreover, the
```

I'm not exactly sure what "losing control" means here. I was expecting something more along the lines of increased chance for bugs or security holes by weakening the type checking because of the unnecessary abilities.

bob-wilson reviewed (approved):
Looks great. Thanks for all the changes!

nilehmann commented:
/land
